### PR TITLE
refactor: switch to NarrowedContext

### PR DESCRIPTION
**Pull request by EdJoPaTo**
Telegraf does ship with `NarrowedContext` with 4.1 which should improve the typings.

The typescript compiler does not seem to be happy about this. Any ideas how to do that better @wojpawlik ?

### CODE REVIEW

File: source/identifier.ts [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
 const BASE_URL = 'http://t.me/#'
 const URL_SEPERATOR = '#'
 
-// TODO: refactor when GuardedContext (or whatever its named) is out (see https://github.com/telegraf/telegraf/discussions/1284)
+// ReplyMessage from typegram is not exported
 type ReplyToMessage = NonNullable<Message.CommonMessage['reply_to_message']>
-export type ReplyToMessageContext<Context extends TelegrafContext> = Context & {message: Message.CommonMessage & {reply_to_message: ReplyToMessage}}
+
+type MessageUpdateContext<Context extends TelegrafContext> = NarrowedContext<Context, Update.MessageUpdate>
+export type ReplyToMessageContext<Context extends TelegrafContext> = MessageUpdateContext<Context> & {message: MessageUpdateContext<Context>['message'] & {reply_to_message: ReplyToMessage}}
```
Comment (wojpawlik) [inline source/identifier.ts:13]:
```suggestion
type NewNonChannelMessage = Update.MessageUpdate['message']
export type ReplyToMessageContext<C extends TelegrafContext> = NarrowedContext<
	C,
	{
		update_id: number
		message: Pick<NewNonChannelMessage, Message.TextMessage> & {
			reply_to_message: ReplyToMessage
		}
	}
>
```

Trickier than I'd like. https://github.com/KnorpelSenf/typegram/pull/10 would have allowed `Update.MessageUpdate<Message.TextMessage & { reply_to_message: ReplyToMessage }>` @KnorpelSenf. Alternatively,

```ts
type UpdateMap = {
  [T in UpdateType]: Extract<tg.Update, Record<T, object>>[T]
}

type NarrowedUpdate<T extends UpdateType, Q, E = unknown> = T extends unknown
  ? { update_id: number } & Record<T, Extract<UpdateMap[T], Q> & E>
  : never
```

allows

```ts
NarrowedUpdate<
  'message',
  Message.TextMessage,
  { reply_to_message: ReplyToMessage }
>
```

Comment (wojpawlik) [reply source/identifier.ts:13]: `NarrowedContext` is actually pretty nasty internally, I'd suggest to wait until I'll try cleaning it up for 4.3...

Comment (wojpawlik) [reply source/identifier.ts:13]: Simpler interface, needs more testing and is too slow for current `NarrowedContext`:

```ts
type OmitUndefined<T extends object, B = T> = T extends unknown
  ? {
      [K in keyof T as B extends Partial<Record<K, undefined>>
        ? never
        : K]: T[K]
    }
  : never

type NarrowedUpdate<
  T extends tt.UpdateType,
  Q extends object
> = T extends unknown
  ? { update_id: number } & Record<
      T,
      OmitUndefined<Deunionize<UpdateMap[T]> & Q>
    >
  : never
```

```ts
NarrowedUpdate<
  'message',
  { text: string; reply_to_message: ReplyToMessage }
>
```

**Comment by EdJoPaTo** [comment]
the current code works so I close this experiment (for now)

**Pull request closed without merging**